### ISSUE TITLE
Deprecate pydot?

### ISSUE THREAD

**Issue by jarrodmillman** [networkx/networkx#5723]
In https://github.com/networkx/networkx/pull/5721, we added a `PendingDeprecationWarning` for `nx.nx_pydot.*`. It has a number of issues and isn't actively maintained. We are also planning to improve installation issues for pygraphviz. This is a tracking issue to gather feedback from users. 

Please leave a comment if you have feelings either way about this change.

**Comment by cans** [comment]
> We are also planning to improve installation issues for pygraphviz.

Do you mean providing wheel package for pygraphviz? I know I had colleagues having trouble building/installing the package on MacOS. That would be awesome!

**Comment by naught101** [comment]
I looked at the PR mentioned, and it looks like the deprecation warning for `nx.nx_pydot.pydot_layout` doesn't suggest a replacement. Is there a way to use pygraphviz via networkx for generating graph layouts? If not, please don't deprecate yet - those layout generators are seriously useful (e.g. the spring layouts are much better than the native networkx ones, as far as I can tell).

**Comment by naught101** [comment]
Ah, I see `nx.nx_agraph.pygraphviz_layout` exists - perhaps those could be added to the deprecation message?

**Comment by me-kell** [comment]
Will `DOT` export further be supported e.g. via `pygraphviz`?

**Comment by roel-v** [comment]
On Windows, it's really hard to get pygraphviz to work. I very much prefer a lightweight tool that only outputs a Graphviz dotfile and then lets me run the Graphviz binary from a script to render diagrams from it myself. I'm not sure what the issues with pydot are, but from my perspective as a user, pygraphviz and pydot may overlap somewhat but still fulfill two different needs. If pydot support actually gets removed I'd be SOL wrt network visualization.

**Comment by dschult** [comment]
Thanks for this comment!
As you may have gathered, our concern with pydot is that it doesn't seem to be maintained anymore. We've started to have bug reports that might just never be fixed.  We've been working pretty hard on getting pygraphviz to be easier to install on Windows. But there are understandable limitations when using a program with such a storied history like Graphviz.

Our support for pydot depends on comments like this one. So thank you.

**Comment by goyalyashpal** [comment]
umh, i just want to render my graph from python script to terminal as inline image, how to do that?

My trail for reaching here:

* DeprecationWarning: nx.nx_pydot.to_pydot depends on the pydot package, which hasknown issues and is not actively maintained. See https://github.com/networkx/networkx/issues/5723

* https://github.com/pydot/pydot/blob/90936e75462c7b0e4bb16d97c1ae7efdf04e895c/README.md#L116
<-- https://www.graphviz.org/resources/#python
<-- https://networkx.org/documentation/stable/reference/drawing.html
<-- https://www.geeksforgeeks.org/visualize-graphs-in-python/

* https://pypi.org/project/imgcat/
<-- ddg 🔍 : [python iterm2 image protocol](https://duckduckgo.com/?q=python+iterm2+image+protocol&t=vivaldi&ia=web)

----

> [NX's] main goal is to enable graph analysis **rather than perform graph visualization**. In the future, graph visualization functionality **may be removed** from NetworkX or only available as an add-on package. 
Proper graph visualization is hard, and we highly recommend that people visualize their graphs **with tools dedicated** to that task.
> \- [NX/Reference/Drawing](https://networkx.org/documentation/stable/reference/drawing.html)

**Comment by goyalyashpal** [comment]
_hiding myself as offtopic_

----
> I'd be SOL wrt network visualization.

SOL = Simply Out of Luck?

https://www.acronymfinder.com/Slang/SOL.html

**Comment by goyalyashpal** [comment]
> we have ascii representations of graphs #5602

the text based thing is nice, i wanted to experiment with that too. thanks for sharing it.



> Are you saying you want to render the graph in the terminal window?

nope, i was saying that 
- to render an image file from graph (i.e. static image, not in a separate window like matplotlib.plot.show() which opens an interactive widget
- the rest of the part is covered thereafter by `imgcat` module which you don't have to worry about 😃

**Comment by goyalyashpal** [comment]
oh, i missed sharing one part of why i dabbled into weeds of this `networkx.drawing.nx_pydot.to_pydot(my_networkx_graph)` method - is that the `ms-python.vscode-pylance` (python LS for VSCode) is showing errors with networkx.draw related functions (wait, i am getting the details of it)

**Comment by dschult** [comment]
I believe matplotlib can save the figure without drawing it interactively.
Maybe use `matplotlib.pyplot.savefig` instead of `matplotlib.pyplot.save`.  There may be other ways too. They definitely allow image file creation without any interactive UI.

**Comment by fzzylogic** [comment]
Just to put a good word in for leaving pydot around for a while, it works on in-browser systems like Jupyter Lite, whereas pygraphviz would not be possible to use there without additional effort compiling a [WASM version](https://github.com/pygraphviz/pygraphviz/issues/319).

**Comment by rossbar** [comment]
> it works on in-browser systems like Jupyter Lite, whereas pygraphviz would not be possible

This is true since `pydot` is pure Python, though it's worth noting that:
 1) You'd only be able to produce dot files - converting to actual image formats (e.g `.png`) would require graphviz to be installed (not sure how difficult this is for in-browser systems), and
 2) It "works" in the sense that it is installable, but there are many defects related to updates to `pyparsing` that pydot never incorporated as it has been unmaintained for nearly 2 years.

**Comment by lkk7** [comment]
@jarrodmillman  @MridulS  A new pydot version has been released which should fix the breaking issues!

**Comment by rossbar** [comment]
Thanks @lkk7 , that's great news! I'm very much in favor of keeping NX's interface to pydot around now that it's actively maintained and working again. I went ahead and tested `nx_pydot` against pydot 2.0 and indeed the original issues we had seen related to pyparsing 3.0 release have been resolved.

I opened #7204 to propose un-deprecating `nx_pydot`. Thanks for the heads up and the latest round of maintenance!
